File: src/AutoContext.Framework/Workers/WorkerHostOptions.cs
namespace AutoContext.Framework.Workers;

/// <summary>
/// Strongly-typed options for the shared worker host: the named pipe to
/// listen on and the stderr ready-marker emitted once the pipe is accepting
/// connections.
/// </summary>
/// <remarks>
/// Each <c>AutoContext.Worker.*</c> process binds and registers an instance
/// of this type. <see cref="Pipe"/> is computed by
/// <c>WorkerHostBuilderExtensions.ConfigureWorkerHost</c> by formatting
/// <c>worker-&lt;workerId&gt;</c> with the parsed
/// <c>--instance-id</c> via
/// <see cref="ServiceAddressFormatter.Format"/>; workers no longer
/// receive an explicit <c>--pipe</c> switch. <see cref="ReadyMarker"/>
/// is supplied by the process itself (e.g.
/// <c>"[AutoContext.Worker.DotNet] Ready."</c>) and is scraped from the
/// worker's stderr by the parent process to detect that the pipe is up.
/// </remarks>
public sealed class WorkerHostOptions
{
    /// <summary>
    /// Named pipe the worker listens on for per-task requests. Computed
    /// at host-bootstrap time from <c>worker-&lt;workerId&gt;</c> +
    /// <c>--instance-id</c>; never set directly via configuration.
    /// </summary>
    public string Pipe { get; init; } = string.Empty;

    /// <summary>
    /// Exact text written to stderr once the pipe server is accepting
    /// connections. Used by parent processes as a readiness handshake.
    /// </summary>
    public string ReadyMarker { get; init; } = string.Empty;

    /// <summary>
    /// Optional service address (<c>autocontext.log#&lt;instance-id&gt;</c>)
    /// the worker connects to for streaming structured log records
    /// (NDJSON) to the parent process. When empty (standalone runs, or
    /// the parent did not pass <c>--service log=...</c>) log output
    /// falls back to stderr.
    /// </summary>
    public string LogServiceAddress { get; init; } = string.Empty;

    /// <summary>
    /// Optional service address
    /// (<c>autocontext.health-monitor#&lt;instance-id&gt;</c>) the worker
    /// connects to on startup to announce its presence to the
    /// extension-side <c>HealthMonitorServer</c>. The connection is held
    /// open for the lifetime of the worker process and the server uses
    /// its closure as the worker's exit signal. When empty (standalone
    /// runs, or the parent did not pass
    /// <c>--service health-monitor=...</c>) no liveness signal is sent.
    /// </summary>
    public string HealthMonitorServiceAddress { get; init; } = string.Empty;
}

File: src/AutoContext.Worker.Web/src/types/worker-host-options.ts
/**
 * Strongly-typed options for a worker host: the named pipe to listen on
 * and the stderr ready-marker emitted once the pipe is accepting
 * connections.
 *
 * Each `AutoContext.Worker.*` process supplies an instance. `pipe`
 * is computed from the `--instance-id` command-line argument and the
 * worker's compile-time worker id.
 * `readyMarker` is process-specific (e.g. `[AutoContext.Worker.Web] Ready.`)
 * and is scraped from the worker's stderr by the parent process to detect
 * that the pipe is up.
 */
export interface WorkerHostOptions {
    /**
     * Named pipe the worker listens on for per-task requests. May be a
     * bare name (`autocontext.web-worker`) or a fully-qualified path;
     * the service normalizes bare names to `\\.\pipe\<name>` on Windows.
     */
    readonly pipe: string;

    /**
     * Exact text written to stderr once the pipe server is accepting
     * connections. Used by parent processes as a readiness handshake.
     */
    readonly readyMarker: string;

    /**
     * Optional named pipe the worker connects to for streaming
     * structured log records (NDJSON) to the parent process. When
     * omitted (the worker is launched standalone, or the parent did
     * not pass `--service log=<address>`), log output falls back to stderr.
     */
    readonly logPipe?: string;

    /**
     * Identifier sent on the LogServer greeting so the extension can
     * route this worker's records to a per-worker output channel
     * (e.g. `AutoContext.Worker.Web`). Counterpart of
     * `IHostEnvironment.ApplicationName` on the .NET side.
     */
    readonly clientName?: string;
}
